p2p/src/peer_manager/peerdb/mod.rs

```diff
@@ -16,197 +16,237 @@
 //! Peer database
 //!
 //! The peer database stores:
-//! - connected peers
-//! - available (discovered) addresses
+//! - all outbound peer addresses
 //! - banned addresses
 //!
 //! Connected peers are those peers that the [`crate::peer_manager::PeerManager`] has an active
 //! connection with. Available addresses are discovered through various peer discovery mechanisms and they are
 //! used by [`crate::peer_manager::PeerManager::heartbeat()`] to establish new outbound connections
 //! if the actual number of active connection is less than the desired number of connections.
 
+pub mod address_data;
 pub mod storage;
 pub mod storage_impl;
+mod storage_load;
 
 use std::{
-    collections::{BTreeMap, BTreeSet},
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    str::FromStr,
     sync::Arc,
     time::Duration,
 };
 
 use common::time_getter::TimeGetter;
-use crypto::random::{make_pseudo_rng, SliceRandom};
+use crypto::random::{make_pseudo_rng, seq::IteratorRandom};
+use logging::log;
 
-use crate::{
-    config,
-    error::{ConversionError, P2pError},
-    net::{AsBannableAddress, NetworkingService},
-};
+use crate::{config, error::P2pError, net::AsBannableAddress};
 
-use self::storage::{
-    PeerDbStorage, PeerDbStorageRead, PeerDbStorageWrite, PeerDbTransactionRo, PeerDbTransactionRw,
+use self::{
+    address_data::{AddressData, AddressStateTransitionTo},
+    storage::{PeerDbStorage, PeerDbStorageWrite},
+    storage_load::LoadedStorage,
 };
 
-pub struct PeerDb<T: NetworkingService, S> {
+use super::MAX_OUTBOUND_CONNECTIONS;
+
+pub struct PeerDb<A, B, S> {
     /// P2P configuration
     p2p_config: Arc<config::P2pConfig>,
 
-    /// Set of currently connected addresses
-    connected_addresses: BTreeSet<T::Address>,
-
-    /// Set of all known addresses
-    known_addresses: BTreeSet<T::Address>,
+    /// Map of all outbound peer addresses
+    addresses: BTreeMap<A, AddressData>,
 
     /// Banned addresses along with the duration of the ban.
     ///
     /// The duration represents the `UNIX_EPOCH + duration` time point, so the ban should end
     /// when `current_time > ban_duration`.
-    banned_addresses: BTreeMap<T::BannableAddress, Duration>,
+    banned_addresses: BTreeMap<B, Duration>,
 
     time_getter: TimeGetter,
 
     storage: S,
 }
 
-impl<T: NetworkingService, S: PeerDbStorage> PeerDb<T, S> {
+impl<
+        A: Ord + FromStr + ToString + Clone + AsBannableAddress<BannableAddress = B>,
```

Comment:
When trait-bounds get so long, using `where` is recommended.

Reply:
Updated, thanks for the hint.